### PR TITLE
Merge feature/vector tiles branch to master

### PR DESCRIPTION
**Pull request by imotov**
We would like to continue developing the vector tile feature under the feature flag in master. We realized that the further development requires refactoring the master code and it would be much easier to do if we have everything in the same branch.

Relates #58696

### CODE REVIEW

File: x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/feature/FeatureFactoryUtils.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Gets the {@link org.elasticsearch.geometry.Geometry} envelope for z/x/y/ tile
+     * in spherical mercator projection.
+     */
+    public static Rectangle getTileBounds(int z, int x, int y) {
+        return new Rectangle(getLong(x, z), getLong(x + 1, z), getLat(y, z), getLat(y + 1, z));
+    }
+
+    private static double getLong(int x, int zoom) {
+        return lonToSphericalMercator(x / Math.pow(2, zoom) * 360 - 180);
```
Comment (jpountz) [inline x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/feature/FeatureFactoryUtils.java:38]: Ss this code performance-sensitive? If so we might want to do leverage `Math#scalb` instead:
```suggestion
        return lonToSphericalMercator(Math.scalb(x, -zoom) * 360 - 180);
```

File: x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/feature/SimpleFeatureFactory.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class SimpleFeatureFactory {
+
+    private final int extent;
+    private final Rectangle rectangle;
+    private final double pointXScale, pointYScale;
+
+    public SimpleFeatureFactory(int z, int x, int y, int extent) {
+        this.extent = extent;
+        rectangle = FeatureFactoryUtils.getTileBounds(z, x, y);
+        pointXScale = 1d / ((rectangle.getMaxLon() - rectangle.getMinLon()) / (double) extent);
```
Comment (jpountz) [inline x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/feature/SimpleFeatureFactory.java:29]: is there a reason not to do this simpler expression instead? If so please add a comment. :)
```suggestion
        pointXScale = (double) extent / (rectangle.getMaxLon() - rectangle.getMinLon());
```

**Review by jpountz — APPROVED** [review]
+1 to develop in master with a feature flag rather than using a feature branch. I didn't fully review, only skimmed through the changes, but please don't hold back merging if both @iverase and yourself think that this has good enough quality to be integrated into the master branch.

File: server/src/main/java/org/elasticsearch/common/util/Maps.java [hunk 3]
```diff
@@ -135,7 +137,44 @@
...
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            if (entry.getValue() instanceof Map) {
+                flatMap.putAll(flatten((Map<String, Object>) entry.getValue(), flattenArrays, ordered, prefix + entry.getKey()));
+            } else if (flattenArrays && entry.getValue() instanceof List) {
+                flatMap.putAll(flatten((List<Object>) entry.getValue(), ordered, prefix + entry.getKey()));
+            } else {
+                flatMap.put(prefix + entry.getKey(), entry.getValue());
+            }
+        }
+        return flatMap;
```
[the remark this answers is not in the document]
Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/util/Maps.java:160]: I think for our purpose overriding is acceptable. We can fix check for it and throw an exception if needed. I will add javadoc with the explanation of the limitations for now.

**Comment by imotov** [comment]
@elasticmachine update branch

**Pull request merged**